design/keycloak.x/operator.md

```diff
@@ -0,0 +1,128 @@
+# Keycloak.X Operator
+
+## Motivation for a new operator
+
+The current Operator made in Go Lang served us and the community well so far, but increasing challenges are paving the road for a complete re-write.
+
+* The codebase is hard to maintain because of organic growth and accumulated technical debt
+* The Keycloak community is more keen to Java and there is less Go Lang expertise
+* The current project needs some high-cost maintenance tasks to be performed in order to use the latest features (e.g. webhooks) and receive the latest fixes and patches, specifically:
+    * upgrading Go lang version from 1.13 to 1.17
+    * upgrading the Operator SDK and the dependencies
+
+  Those upgrades will require creating a completely new project, using different libraries, moving, and in some cases, rewriting the components, e.g. the whole testsuite.
+* The current approach around CRDs no longer fits the long term vision for cloud-native Keycloak as it is very error-prone and fragile.
+* A Java operator can share business objects with the Keycloak main codebase increasing the code-reuse and dramatically reducing the chances of introducing bugs in the translation process from Kubernetes resources.
+* A unified codebase will make it easy to test and debug the entire system.
+* A new operator will embrace from the ground up the new Cloud Native capabilities of upcoming Keycloak releases such as the Quarkus distribution and Store.X, making those first-class citizen overall improving the user experience.
+
+
+## Features
+
+---
+**NOTE**
+
+The primary target of the operator is to make it easy to achieve production grade installations of Keycloak.X.
+
+---
+
+### Configuring Keycloak deployment
+
+The operator will use a CRD representing the Keycloak installation. The CRD will expose the following configuration options:
+* custom Keycloak image; default base image will be used if not specified
+* Git configuration to fetch static business objects configuration (see below)
+* manual horizontal and vertical scaling
+* pod and node affinity, toleration
+* SSL config, truststore
```

Comment:
We support different proxy modes, some of them should not require TLS (e.g.: edge). But I guess we always want a secure channel between Keycloak and the ingress/route, even if running within the cluster network. Perhaps this requirement should be more explicit?

In addition to that, do we want to enable HTTP? Is this somewhat needed by Grafana (or has a complex set up) or any other client/service that needs to communicate with Keycloak?

I'm also missing exposing options to manage some key ingress/route settings. I think these are very common settings you usually want to change and have some control:

* connection limit
* timeout
* LB strategy

Reply:
We'll add more details around SSL.

The current plan is that the operator will provide only default Ingress setting. If you want to control Ingress and configure it, you can do it directly. You can even deploy a custom Ingress.